### PR TITLE
Update Coalesce function semantics to short circuit

### PR DESCRIPTION
**Pull request by gregli-msft**
Fixes https://github.com/microsoft/Power-Fx/issues/1831 and https://github.com/microsoft/Power-Fx/issues/1066.  Added additional tests for short circuiting.  We also need to be careful not to break Canvas semantics before Power Fx 1.0 is adopted there.

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by CarlosFigueiraMSFT** [comment]
>     public class TxtFileDataAttribute : DataAttribute

Do we still need this attribute with the new one?

---
Refers to: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestHelpers/TxtFileDataAttribute.cs:23 in 3ad4d0a. [](commit_id = 3ad4d0a94a331e22dd6f902efd7873ff057af5a6, deletion_comment = False)

**Comment by CarlosFigueiraMSFT** [comment]
>         }

Can we remove this virtual? This would need to be updated in Power Apps as well, but the changes are fairly small there.

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs:571 in 3ad4d0a. [](commit_id = 3ad4d0a94a331e22dd6f902efd7873ff057af5a6, deletion_comment = False)

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestHelpers/TxtFileDataAttribute.cs [hunk 1]
```diff
@@ -97,5 +97,59 @@ internal static string GetDefaultTestDir(string filePath)
...
+                {
+                    list.Add(new object[] { file.Name });
+                }
+            }
+            catch (Exception e)
+            {
+                // If this method crashes, then we just get 0 tests. 
+                // The only way to communicate a failure from here back to the developer
+                // is to pass a "fake" test object that always fails and contains the error.
+                // In this ecase, this will be a bogus file name that won't load.
```
[the remark this answers is not in the document]
Comment (gregli-msft) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestHelpers/TxtFileDataAttribute.cs:145]: Fixed.  Thanks!

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by CarlosFigueiraMSFT — APPROVED** [review]
🚢

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**